File: src/main/java/com/cjcrafter/openai/chat/ChatResponse.java
package com.cjcrafter.openai.chat;

import com.google.gson.JsonObject;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * This is the object returned from the API. You want to access choices[0]
 * to get your response.
 */
public final class ChatResponse {

    private final String id;
    private final String object;
    private final long created;
    private final List<ChatChoice> choices;
    private final ChatUsage usage;

    public ChatResponse(String id, String object, long created, List<ChatChoice> choices, ChatUsage usage) {
        super();
        this.id = id;
        this.object = object;
        this.created = created;
        this.choices = choices;
        this.usage = usage;
    }

    public ChatResponse(JsonObject json) {
        this(json.get("id").getAsString(), json.get("object").getAsString(), json.get("created").getAsLong(), StreamSupport.stream(json.get("choices").getAsJsonArray().spliterator(), false).map(element -> new ChatChoice(element.getAsJsonObject())).collect(Collectors.toList()), new ChatUsage(json.get("usage").getAsJsonObject()));
    }

    public String getId() {
        return this.id;
    }

    public String getObject() {
        return this.object;
    }

    public long getCreated() {
        return this.created;
    }

    public List<ChatChoice> getChoices() {
        return this.choices;
    }

    public ChatUsage getUsage() {
        return this.usage;
    }
}

File: src/main/kotlin/com/cjcrafter/openai/chat/ChatResponse.kt
package com.cjcrafter.openai.chat

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.*

/**
 * The `ChatResponse` contains all the data returned by the OpenAI Chat API.
 * For most use cases, [ChatResponse.get] (passing 0 to the index argument) is
 * all you need.
 *
 * @property id      The unique id for your request.
 * @property created The Unix timestamp (measured in seconds since 00:00:00 UTC on January 1, 1970) when the API response was created.
 * @property choices The list of generated messages.
 * @property usage   The number of tokens used in this request/response.
 * @constructor Create Chat response (for internal usage).
 */
data class ChatResponse(
    @JsonProperty(required = true) val id: String,
    @JsonProperty(required = true) val created: Long,
    @JsonProperty(required = true) val choices: List<ChatChoice>,
    @JsonProperty(required = true) val usage: ChatUsage
) {

    /**
     * Returns the [Instant] time that the OpenAI Chat API sent this response.
     * The time is measured as a unix timestamp (measured in seconds since
     * 00:00:00 UTC on January 1, 1970).
     *
     * Note that users expect time to be measured in their timezone, so
     * [getZonedTime] is preferred.
     *
     * @return The instant the api created this response.
     * @see getZonedTime
     */
    fun getTime(): Instant {
        return Instant.ofEpochSecond(created)
    }

    /**
     * Returns the time-zoned instant that the OpenAI Chat API sent this
     * response. By default, this method uses the system's timezone.
     *
     * @param timezone The user's timezone.
     * @return The timezone adjusted date time.
     * @see TimeZone.getDefault
     */
    @JvmOverloads
    fun getZonedTime(timezone: ZoneId = TimeZone.getDefault().toZoneId()): ZonedDateTime {
        return ZonedDateTime.ofInstant(getTime(), timezone)
    }

    /**
     * Shorthand for accessing the generated messages (shorthand for
     * [ChatResponse.choices]).
     *
     * @param index The index of the message (`0` for most use cases).
     * @return The generated [ChatChoice] at the index.
     */
    operator fun get(index: Int): ChatChoice {
        return choices[index]
    }
}